New Connector Type  - D3 security SOAR

d3-naiji opened the pull request:
## Summary
Introduces a new built-in action type and sends the ElasticSearch alerts to D3 Security SOAR for incident response, workflow automation and deep investigation. This integrated alert action allows you to create Events or trigger Playbook workflow actions automatically in the D3 Security SOAR platform.

The new action type and the action connector with D3 Security SOAR are as follows:

<img width="224" alt="D3ConnectorCard2" src="https://user-images.githubusercontent.com/109116834/200965642-6fef1a0b-232c-40c2-8556-d5abc50640f5.png">
<img width="721" alt="D3ConnectorForm1" src="https://user-images.githubusercontent.com/109116834/200965662-5b940b68-ba03-41ab-889d-8f7082333b2c.png">

and action params form available in Create Alert form:

<img width="555" alt="D3_new-Form" src="https://user-images.githubusercontent.com/109116834/201801116-6b84fd2a-891c-4b88-af04-9d193efc7d0a.png">

OBS.: merge conflict has been resolved and new fields have been added for params form.


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)



### For maintainers

(https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

TinaHeiligers commented:
buildkite test this

TinaHeiligers commented:
@elasticmachine merge upstream

TinaHeiligers commented:
@elasticmachine merge upstream

TinaHeiligers commented on x-pack/plugins/stack_connectors/server/connector_types/stack/d3security/index.ts:
```diff
@@ -0,0 +1,338 @@
+  const { body:data, severity, eventType } = params;
+
+  const axiosInstance = axios.create();
+  console.log(data);
```
not allowed, please remove or use suitable log mocking

TinaHeiligers commented on x-pack/plugins/stack_connectors/server/connector_types/stack/d3security/index.ts:
```diff
@@ -0,0 +1,338 @@
+    bodyObj['hits'] = {};
+    bodyObj['hits']['hits'] = {};
+    bodyObj['hits']['hits']["_source"] = {}
+    console.log("raw", rawdata);
```
not allowed, please remove or use suitable log mocking

TinaHeiligers reviewed (changes requested):
initial build failures:
`console.log` not allowed, please remove or use suitable log mocking

TinaHeiligers commented:
@elastic/response-ops could you please take a look at the proposal?

d3-naiji commented:
> initial build failures: `console.log` not allowed, please remove or use suitable log mocking

Hi @TinaHeiligers, we have already removed 'console.log' in the latest commits. Please, review it again. Thank you!

d3-naiji commented:
@elasticmachine merge upstream

TinaHeiligers commented:
@elastic/response-ops could you please take a look at the proposal?

d3-naiji commented:
@TinaHeiligers  Hi Tina, any update from the code review? Is there anything we can help with?

TinaHeiligers commented:
@d3-naiji I am removing myself as a reviewer to allow the relevant team to take this further.

TinaHeiligers reviewed (commented):
CC @elastic/response-ops

TinaHeiligers reviewed (commented):
I cannot comment on the code, nor am I able to remove myself as a reviewer, hence a comment as such, rather than an explicit review.

d3-naiji commented:
Thanks @TinaHeiligers  

@elastic/response-ops  Please let me know if there is any feedback or changes. Thanks

mikecote commented:
Hi @d3-naiji, what is the difference between this PR and https://github.com/elastic/kibana/pull/137362? I see @semd has been active in reviewing the other PR and I also see `-v2` appended to the branch name of this PR. Happy to have @semd start a review of this one if still needed (cc @stephmilovic).

d3-naiji commented:
> Hi @d3-naiji, what is the difference between this PR and #137362? I see @semd has been active in reviewing the other PR and I also see `-v2` appended to the branch name of this PR. Happy to have @semd start a review of this one if still needed (cc @stephmilovic).

Hi @mikecote @semd @stephmilovic ,
This PR is a new branch that we merged the latest codes from Main, and added a few changes to the UI.
Because there were a lot of file structure changes between the feature branch in #137362 and the latest Main. It was hard to resolve all the conflicts, so we created a new branch based on the latest Main and implemented the features again.

semd commented:
Hi @d3-naiji, 
I see there have been changes in the _connector_types_ folders, there's no longer any `/stack` directory. Could you move the connector directories:
`x-pack/plugins/stack_connectors/public/connector_types/stack/d3security` -> `x-pack/plugins/stack_connectors/public/connector_types/d3security`
and the server
`x-pack/plugins/stack_connectors/server/connector_types/stack/d3security` -> `x-pack/plugins/stack_connectors/server/connector_types/d3security`

We'll start testing it soon, but by now we can start with this, after doing the relocation it should be easier to merge `main` again.

semd commented:
@d3-naiji I would also like to confirm that this new connector will only be needed for Detection Alerts. It will not be available for Cases. Is this correct?

d3-naiji commented:
> @d3-naiji I would also like to confirm that this new connector will only be needed for Detection Alerts. It will not be available for Cases. Is this correct?

Yes, correct.

d3-naiji commented:
> Hi @d3-naiji, I see there have been changes in the _connector_types_ folders, there's no longer any `/stack` directory. Could you move the connector directories: `x-pack/plugins/stack_connectors/public/connector_types/stack/d3security` -> `x-pack/plugins/stack_connectors/public/connector_types/d3security` and the server `x-pack/plugins/stack_connectors/server/connector_types/stack/d3security` -> `x-pack/plugins/stack_connectors/server/connector_types/d3security`
> 
> We'll start testing it soon, but by now we can start with this, after doing the relocation it should be easier to merge `main` again.

Thank you for the update. I will make the necessary changes and inform you once it is done.

d3-naiji commented:
Hi @semd , Just wanted to let you know that I've updated the branch and merged all the commits with the latest main. Could you please take a look and let me know if everything's good for testing? And if you need any further changes from my team, feel free to contact me anytime.

semd commented:
buildkite test this

semd commented:
@d3-naiji There are some checks and tests failing. Do you have access to see the details?

d3-naiji commented:
> @d3-naiji There are some checks and tests failing. Do you have access to see the details?

Yes, I can see the details in this link https://buildkite.com/elastic/kibana-pull-request/builds/87307
I will look into it.

d3-bensong commented:
@elasticmachine merge upstream

stephmilovic commented:
@elasticmachine merge upstream

d3-bensong commented:
[@elasticmachine](https://github.com/elasticmachine) merge upstream

cnasikas commented:
@elasticmachine merge upstream

cnasikas commented:
Hey @d3-bensong! Could you please address the CLA issue? Thanks!

d3-bensong commented:
[@elasticmachine](https://github.com/elasticmachine) merge upstream

cnasikas commented:
@elasticmachine merge upstream

cnasikas commented:
buildkite test this

cnasikas commented:
@elasticmachine, run elasticsearch-ci/docs

cnasikas commented:
@d3-naiji It seems that the CLA is fixed now. For some weird reason, I cannot merge with upstream. Could you please do it manually?

cnasikas commented:
@d3-naiji @d3-bensong Is there any plan to integrate the connector with Cases?

d3-bensong commented:
@elasticmachine merge upstream

d3-naiji commented:
> @d3-naiji @d3-bensong Is there any plan to integrate the connector with Cases?

Hi @cnasikas , 
We would like to enhance it in the future to support "Case".

stephmilovic commented:
<img width="666" alt="Screenshot 2023-05-26 at 12 07 25 PM" src="https://github.com/elastic/kibana/assets/6935300/0b0125ad-2e84-4f91-af90-a635491b0bd8">
Is this intended?

semd commented:
buildkite test this

semd commented:
@elasticmachine, run elasticsearch-ci/docs

d3-bensong commented:
@elasticmachine merge upstream

d3-bensong commented:
@elasticmachine, run elasticsearch-ci/docs

d3-bensong commented:
Hi @cnasikas ,
We fixed problem, can you try again please?

cnasikas commented:
buildkite test this

cnasikas commented:
@elasticmachine, run elasticsearch-ci/docs

d3-bensong commented:
Hi @cnasikas, 
I missed an issue last time, now fixed, could you try it again please, thanks.

stephmilovic commented:
Hello @d3-bensong and @d3-naiji. After review, because of the planned cases enhancements it was determined that the D3 Security connector needed to be transitioned to our sub actions framework. In order to move this PR forward in time for our upcoming 8.9 Feature Freeze date, we determined it would be most efficient if we move forward with an Elastic owned branch and have our team carry out the rest of the work. There is a [new PR open](https://github.com/elastic/kibana/pull/158569) with the sub actions refactor complete, along with additional required tests and documentation. We are moving forward with a review on that PR today. Your work is included in that PR. Therefore, you can go ahead and close this one. I will follow up over email to confirm once I have the new PR merged. Let me know if you have any questions.

Thank you!

d3-naiji commented:
@stephmilovic  Thanks! I will close this PR
